libcudacxx/include/cuda/std/__execution/env.h

```diff
@@ -0,0 +1,319 @@
+_CCCL_HOST_DEVICE prop(_Query, _Value) -> prop<_Query, _Value>;
+
+/**
+ * @brief A variadic template structure representing an environment.
+ *
+ * This structure encapsulates a tuple of environments and provides functionality to query
+ * the first environment that satisfies a given query type.
+ *
+ * @tparam _Envs... Variadic template parameter pack representing the types of
+ * environments.
+ */
+template <class... _Envs>
+struct _CCCL_TYPE_VISIBILITY_DEFAULT env
+{
+  /**
+   * @brief Retrieves the first environment that satisfies the given query type.
+   *
+   * This static constexpr function checks each environment in the tuple to find
+   * the first one that is queryable with the specified query type. If such an
+   * environment is found, it is returned.
+   *
+   * @tparam _Query The type of the query to be performed.
+   * @param __self A constant reference to the current `env` instance.
+   * @return The first environment in the tuple that satisfies the query type.
+   * @note If no environment satisfies the query, the behavior is undefined.
+   */
+  template <class _Query>
+  _CCCL_TRIVIAL_API static constexpr decltype(auto) __get_1st(const env& __self) noexcept
+  {
+    // NOLINTNEXTLINE (modernize-avoid-c-arrays)
+    constexpr bool __flags[] = {__queryable_with<_Envs, _Query>..., false};
+    constexpr size_t __idx   = detail::__find_pos(__flags, __flags + sizeof...(_Envs));
+    if constexpr (__idx != detail::__npos)
+    {
+      return __cget<__idx>(__self.__envs_);
+    }
+  }
+
+  /**
+   * @brief Alias for the type of the first environment that satisfies the query type.
+   *
+   * This alias resolves to the type of the first environment in the tuple that
+   * satisfies the specified query type.
+   *
+   * @tparam _Query The type of the query to be performed.
+   */
+  template <class _Query>
+  using __1st_env_t _CCCL_NODEBUG_ALIAS = decltype(env::__get_1st<_Query>(declval<const env&>()));
+
+  /**
+   * @brief Queries the first environment that satisfies the given query type.
+   *
+   * This function invokes the `query` method on the first environment in the tuple
+   * that satisfies the specified query type, passing the query object as an argument.
+   *
+   * @tparam _Query The type of the query to be performed.
+   * @param __query The query object to be passed to the environment's `query` method.
+   * @return The result of the `query` method on the first environment that satisfies the query type.
+   * @throws noexcept If the query operation is noexcept for the resolved environment and query type.
+   */
+  _CCCL_TEMPLATE(class _Query)
+  _CCCL_REQUIRES(__queryable_with<__1st_env_t<_Query>, _Query>)
+  _CCCL_TRIVIAL_API constexpr auto query(_Query __query) const
+    noexcept(__nothrow_queryable_with<__1st_env_t<_Query>, _Query>) -> __query_result_t<__1st_env_t<_Query>, _Query>
+  {
+    return env::__get_1st<_Query>(*this).query(__query);
+  }
+
+  __tuple<_Envs...> __envs_;
```

Comment:
This is inconsistent with the specialization below, as we are missing
```suggestion
  _CCCL_NO_UNIQUE_ADDRESS __tuple<_Envs...> __envs_;
```